core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -0,0 +1,254 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import io.grpc.Status;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Transports for a single server.
+ */
+@ThreadSafe
+final class TransportSet {
+  private static final Logger log = Logger.getLogger(TransportSet.class.getName());
+
+  private final Object lock = new Object();
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+  private final Callback callback;
+  private final ClientTransportFactory transportFactory;
+  private final ScheduledExecutorService scheduledExecutor;
+
+  @GuardedBy("lock")
+  @Nullable
+  private BackoffPolicy reconnectPolicy;
+
+  @GuardedBy("lock")
+  @Nullable
+  private ScheduledFuture<?> reconnectTask;
+
+  /**
+   * All transports that are not stopped. At the very least the value of {@link
+   * activeTransportFuture} will be present, but previously used transports that still have streams
+   * or are stopping may also be present.
+   */
+  @GuardedBy("lock")
+  private final Collection<ClientTransport> transports = new ArrayList<ClientTransport>();
+
+  @GuardedBy("lock")
+  private boolean shutdown;
+
+  /**
+   * The future for the transport for new outgoing requests. 'lock' lock must be held when assigning
+   * to activeTransportFuture.
+   */
+  @GuardedBy("lock")
+  @Nullable
+  private volatile SettableFuture<ClientTransport> activeTransportFuture;
+
+  TransportSet(BackoffPolicy.Provider backoffPolicyProvider,
+      ClientTransportFactory transportFactory, ScheduledExecutorService scheduledExecutor,
+      Callback callback) {
+    this.backoffPolicyProvider = backoffPolicyProvider;
+    this.transportFactory = transportFactory;
+    this.scheduledExecutor = scheduledExecutor;
+    this.callback = callback;
+    connectActiveTransport();
+  }
+
+  /**
+   * Returns a future for the active transport that will be used to create new streams.
+   *
+   * <p>Depending on timing, if this {@code TransportSet} has been shut down, either this method or
```

Comment:
I think this is fine, especially for now, but it could be reasonable to never return `null` here and just let the future's value be `null`. It would be very easy to return a (maybe static) `Futures.immediateFuture(null)` when shutdown.